File: bms/go/geo/geo_s2_CellIDFromString/geo_s2_CellIDFromString.go
package test

const (
	geo_s2_MaxLevel = 30
	geo_s2_PosBits  = 2*geo_s2_MaxLevel + 1
)

func geo_s2_CellIDFromString(s string) geo_s2_CellID {
	level := len(s) - 2
	if level < 0 || level > geo_s2_MaxLevel {
		return geo_s2_CellID(0)
	}
	face := int(s[0] - '0')
	if face < 0 || face > 5 || s[1] != '/' {
		return geo_s2_CellID(0)
	}
	id := geo_s2_CellIDFromFace(face)
	for i := 2; i < len(s); i++ {
		childPos := s[i] - '0'
		if childPos < 0 || childPos > 3 {
			return geo_s2_CellID(0)
		}
		id = id.Children()[childPos]
	}
	return id
}

func geo_s2_CellIDFromFace(face int) geo_s2_CellID {
	return geo_s2_CellID((uint64(face) << geo_s2_PosBits) + geo_s2_lsbForLevel(0))
}

func geo_s2_lsbForLevel(level int) uint64 { return 1 << uint64(2*(geo_s2_MaxLevel-level)) }

func (ci geo_s2_CellID) Children() [4]geo_s2_CellID {
	var ch [4]geo_s2_CellID
	lsb := geo_s2_CellID(ci.lsb())
	ch[0] = ci - lsb + lsb>>2
	lsb >>= 1
	ch[1] = ch[0] + lsb
	ch[2] = ch[1] + lsb
	ch[3] = ch[2] + lsb
	return ch
}

func (ci geo_s2_CellID) lsb() uint64 { return uint64(ci) & -uint64(ci) }

type geo_s2_CellID uint64

File: translations_only/fail/go/geo/geo_s2_CellIDFromString/gpt4/Hinted/geo_s2_CellIDFromString.rs

const GEO_S2_MAX_LEVEL: usize = 30;
const GEO_S2_POS_BITS: u64 = 2 * GEO_S2_MAX_LEVEL as u64 + 1;

#[derive(Clone, Copy)]
struct GeoS2CellID(u64);

fn geo_s2_cell_id_from_string(s: &str) -> GeoS2CellID {
    let level = s.len() - 2;
    if level < 0 || level > GEO_S2_MAX_LEVEL {
        return GeoS2CellID(0);
    }
    let face = s.chars().nth(0).unwrap() as i32 - '0' as i32;
    if face < 0 || face > 5 || s.chars().nth(1).unwrap() != '/' {
        return GeoS2CellID(0);
    }
    let mut id = geo_s2_cell_id_from_face(face as usize);
    for c in s[2..].chars() {
        let child_pos = c as i32 - '0' as i32;
        if child_pos < 0 || child_pos > 3 {
            return GeoS2CellID(0);
        }
        id = children(&id)[child_pos as usize];
    }
    id
}

fn geo_s2_cell_id_from_face(face: usize) -> GeoS2CellID {
    GeoS2CellID((face as u64) << GEO_S2_POS_BITS | geo_s2_lsb_for_level(0))
}

fn geo_s2_lsb_for_level(level: usize) -> u64 {
    1 << (2 * (GEO_S2_MAX_LEVEL - level)) as u64
}

fn children(ci: &GeoS2CellID) -> [GeoS2CellID; 4] {
    let mut ch = [GeoS2CellID(0); 4];
    let lsb = lsb(ci);
    ch[0] = GeoS2CellID(ci.0 - lsb + (lsb >> 2));
    let lsb = lsb >> 1;
    ch[1] = GeoS2CellID(ch[0].0 + lsb);
    ch[2] = GeoS2CellID(ch[1].0 + lsb);
    ch[3] = GeoS2CellID(ch[2].0 + lsb);
    ch
}

fn lsb(ci: &GeoS2CellID) -> u64 {
    ci.0 & ci.0.wrapping_neg()
}

fn main() {
    println!("{:?}", geo_s2_cell_id_from_string("1[").0);
    println!("{:?}", geo_s2_cell_id_from_string("/").0);
    println!("{:?}", geo_s2_cell_id_from_string("3/0123").0);
    println!("{:?}", geo_s2_cell_id_from_string("4/2023").0);
}
